### ISSUE TITLE
Intermitent decryption problem with Kerberos

### ISSUE THREAD

**Issue by vdesjardins** [WinRb/WinRM#161]
Hi,

We have an intermittent problem with some responses from the WinRM daemon. We are using the WinRM gem version 1.3.5 on RHEL 7. What we can see from the exception is that the end of the received message seems corrupted:

```
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/http/response_handler.rb:44:in `rescue in response_xml': Unable to parse WinRM response: #<ArgumentError: invalid byte sequence in UTF-8> (WinRM::WinRMHTTPTransportError)
/home/gen00698/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rexml/source.rb:219:in `match'
/home/gen00698/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rexml/source.rb:219:in `match'
/home/gen00698/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rexml/parsers/baseparser.rb:339:in `pull_event'
/home/gen00698/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rexml/parsers/baseparser.rb:184:in `pull'
/home/gen00698/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rexml/parsers/treeparser.rb:22:in `parse'
/home/gen00698/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rexml/document.rb:287:in `build'
/home/gen00698/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/rexml/document.rb:44:in `initialize'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/http/response_handler.rb:42:in `new'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/http/response_handler.rb:42:in `response_xml'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/http/response_handler.rb:36:in `parse_to_xml'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/http/transport.rb:151:in `send_request'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:430:in `send_message'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:169:in `run_command'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:300:in `block in run_cmd'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:131:in `open_shell'
/home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:299:in `run_cmd'
test_winrm.rb:105:in `main_simple'
test_winrm.rb:111:in `<main>'
...
Exception parsing
Line: 1
Position: 846
Last 80 unconsumed characters:
</s:BodVT<90>?[Õ}<82>=Ô<97>^A¯¯ (200).
        from /home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/http/response_handler.rb:42:in `response_xml'
        from /home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/http/response_handler.rb:36:in `parse_to_xml'
        from /home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/http/transport.rb:151:in `send_request'
        from /home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:430:in `send_message'
        from /home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:169:in `run_command'
        from /home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:300:in `block in run_cmd'
        from /home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:131:in `open_shell'
        from /home/gen00698/.rvm/gems/ruby-2.2.3@deploy-rhel7/gems/winrm-1.3.5/lib/winrm/winrm_service.rb:299:in `run_cmd'
        from test_winrm.rb:105:in `main_simple'
        from test_winrm.rb:111:in `<main>'
```

The command used to test is a simple 'ipconfig /all".

Any idea what can cause this?

Thanks.

**Comment by tricktrap** [comment]
:+1: Me too, but I thought it was on my end. It is usually ok if you retry your connection. It always appears to be the very last batch of the connection, with it happening right as the SOAP body tag is closing. For me, adding a retry to my code if this happens allowed me to work around the issue.

**Comment by sneal** [comment]
Does it work properly on 1.3.4? 1.3.5 forced encoding to UTF-8.

**Comment by tricktrap** [comment]
I'll check that on my end.

**Comment by tricktrap** [comment]
It looks like it still happens with the gem at 1.3.4. I'm running my script over and over again as fast as possible, which appears to be a way to recreate the issue (it only happens maybe 1 in 15-20 times).

**Comment by vdesjardins** [comment]
I am able to reproduce it too with 1.3.4. I'm testing it with a for loop 30 times and got the error between 1 and 3 times per test.

**Comment by mwrock** [comment]
What version of windows is the endpoint running?

**Comment by vdesjardins** [comment]
2008 R2 SP1. WinRM version is 3.0 as per command winrm id:

```
C:\Windows\system32>winrm id
IdentifyResponse
    ProtocolVersion = http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd
    ProductVendor = Microsoft Corporation
    ProductVersion = OS: 6.1.7601 SP: 1.0 Stack: 3.0
    SecurityProfiles
        SecurityProfileName = http://schemas.dmtf.org/wbem/wsman/1/wsman/secprofile/http/spnego-kerberos
```

**Comment by dvucko** [comment]
This is still an issue with WinRb 1.8.1. (Linux Mint), e.g. when targeting Win Server 2012.

**Comment by jerryk55** [comment]
@dvucko can you be more specific with what you mean by "Switching to negotiate seems to be the only known solution so far"?  I believe we are seeing this issue as well.  Thanks much

**Comment by dvucko** [comment]
@jerryk55 the problem only occurs (in a seemingly non-deterministic manner) if you choose Kerberos in your script, i.e.:
`winrm = WinRM::WinRMWebService.new(endpoint, :kerberos, :realm => krb5_realm)`
If you use negotiate authentication, the problem goes away:
`winrm = WinRM::WinRMWebService.new(endpoint, :negotiate, :user => user, :pass => password)`

**Comment by jerryk55** [comment]
@dvucko ah thanks for that.  I guess what I'm seeing is different since I"m always using :negotiate.

**Comment by danielladd** [comment]
I get this on almost every single execution, anyone aware of a work around that does not involve disabling kerberos?

**Comment by aalblas** [comment]
Did anyone come up with with a workaround that doesn't disable kerberos? I'm seeing this approximately 10% of my requests, simply calling the same call again will nearly always work without issue.

**Comment by alv000h** [comment]
+1

It appears to be an exception with parser...

**Comment by Lennie** [comment]
Here is where things go wrong for me:

When I edit lib/winrm/http/transport.rb

And change:
```
        @logger.info "SOAP message decrypted (MAJ: #{maj_stat}, " \
          "MIN: #{min_stat.read_int}):\n#{iov1[:buffer].value}"

```
To:
```

        STDOUT.write "SOAP message decrypted (MAJ: #{maj_stat}, " \
          "MIN: #{min_stat.read_int}):\n#{iov1[:buffer].value}\n"

```

I would normally see:

SOAP message decrypted (MAJ: 0, MIN: 0):

But when things go wrong, I see:

SOAP message decrypted (MAJ: 393216, MIN: -1765328353):

Not sure if that is a problem with decrypting by the GSSAPI-library or maybe a problem with the unpacks.

The result of len = str.unpack('L').first is always 60 in my case.

**Comment by Iristyle** [comment]
I fixed this bug in #302 and it was released in version 2.3.2 of the gem, so this can be closed!
